Lib/glyphsLib/builder/components.py

```diff
@@ -15,10 +15,52 @@
 from __future__ import (print_function, division, absolute_import,
                         unicode_literals)
 
+from .constants import GLYPHS_PREFIX
 
-def to_ufo_draw_components(self, pen, components):
+
+def to_ufo_draw_components(self, ufo_glyph, components):
     """Draw .glyphs components onto a pen, adding them to the parent glyph."""
+    pen = ufo_glyph.getPointPen()
 
     for component in components:
         pen.addComponent(component.name,
                          component.transform)
+
+
+def to_glyphs_draw_components(self, ufo_glyph, layer):
+    for comp in ufo_glyph.components:
+        component = self.glyphs_module.GSComponent(comp.baseGlyph)
+        component.transform = comp.transformation
+        layer.components.append(component)
+
+
+def to_ufo_component_data(self, ufo_glyph, layer):
+    # data related to components stored in lists of booleans
+    # each list's elements correspond to the components in order
+    for key in ['alignment', 'locked']:
+        values = []
+        for c in layer.components:
+            value = getattr(c, key)
+            # FIXME: (jany) bad idea to skip none because it shifts indices
+            # TODO: (jany) switch to a dict structure: {componentId: {data}}
```

Comment:
Why is it necessary to distinguish between the public.glyphOrder and the con.schriftgestaltung.glyphOrder? Isn’t the former enough?

Reply:
Duplicating the glyphOrder is not only confusing but may lead to the two values getting out of sync with each other

Reply:
> Would it make sense for GlyphsApp to use that for UFO3 since components can have identifiers.

I’ll follow any working solution. 

> Why is it necessary to distinguish between the public.glyphOrder and the con.schriftgestaltung.glyphOrder? Isn’t the former enough?

They mean different things. The `public.glyphOrder` stores the order of the glyphs as the happened to be at that moment. `con.schriftgestaltung.glyphOrder` is just the custom parameter, It doesn't need to be in sync with the actual order and can contain names that are not in the font and doesn’t need to have all glyphs. So you should mostly be using `public.glyphOrder` and just keep `con.schriftgestaltung.glyphOrder` around for round tripping.

Reply:
The way you described the com.schriftgestaltung.glyphOrder sounds to me very much like the way the UFO spec defines the public.glyphOrder. The fact is the UFO glyph sets don’t have any “actual order” as contents.plist is just a mapping without any internal order (it’s alphabetically sorted in most implementations but doesn’t imply anything).